### PR TITLE
feat(messages): validate message length before sending

### LINKED ISSUE #139: feat(messages): validate message length before sending to prevent silent Slack truncation
## Problem

Slack silently truncates messages exceeding 40,000 characters. When `slck messages send` posts a message that's too long, Slack returns `ok: true` with a `message_truncated` warning in response metadata — but `slck` doesn't parse that warning, so it reports success. The caller has no idea their message was cut off mid-content.

**Real-world example:** An AI agent used `slck messages send` to post a ~45K character questionnaire into a Slack thread. The message was truncated mid-word ("Extensio...") and the agent never knew.

## Solution

Add client-side message length validation to `messages send` and `messages update`:

1. **Error by default** when `text` exceeds 40,000 characters (Slack's hard limit)
2. **Actionable error message** with character count and alternatives:
   - `--file <path>` to upload as a file attachment
   - `slck canvas create` to create a Slack canvas
   - Manual chunking suggestion
3. **Skip validation** for file uploads (the `--file` path doesn't have this limit)

### Code changes
- `internal/cmd/messages/messages.go` — `maxMessageTextLen` constant + `validateMessageLength()` function
- `internal/cmd/messages/send.go` — validation call in `runSend`
- `internal/cmd/messages/update.go` — validation call in `runUpdate`
- Tests for all of the above

### Documentation
- `CHANGELOG.md` — true up: collapse `[Unreleased]` into versioned sections (`[3.2.0]` for new + post-v3.1.35 work, `[3.1.35]` for previously released items)
- `README.md` — add Message Length Limits entry under Known Limitations

### PR DESCRIPTION
**Pull request by rianjs**
## Summary
- Adds pre-send validation for message text length against Slack's 40,000 character hard limit
- Returns an actionable error with alternatives (`--file`, `canvas create`, manual chunking) instead of letting Slack silently truncate
- Applies to both `messages send` and `messages update`; skipped for file uploads
- Trues up CHANGELOG.md: removes `[Unreleased]`, creates `[3.2.0]` and `[3.1.35]` sections
- Adds Message Length Limits entry to README Known Limitations

Closes #139

## Test plan
- [x] `TestValidateMessageLength` — under/at/over limit boundary tests
- [x] `TestRunSend_MessageTooLong` — send rejects oversized text
- [x] `TestRunUpdate_MessageTooLong` — update rejects oversized text
- [x] `TestRunSend_FileUpload_LongTextAllowed` — file uploads bypass length check
- [x] Manual: `python3 -c "print('x' * 50000)" | ./bin/slck messages send C123 -` → error with helpful message
- [x] `make build && make test && make lint` all pass

### CODE REVIEW

**Comment by rianjs** [comment]
## Test Coverage Review

All four new tests pass. Coverage is solid for a focused validation feature.

**What's covered well:**
- Unit test for `validateMessageLength` hits all meaningful boundary cases: empty, under limit, at exact limit, over by one, well over. Boundary-at-limit (40000) returning `nil` is the most important edge case and it's there.
- Error message content is asserted (`--file`, `canvas create` hints), not just the presence of an error.
- `runSend` integration: long text rejected correctly.
- `runUpdate` integration: long text rejected correctly.
- File-upload bypass: long text with `--file` is explicitly allowed — this matches the skip-validation branch in `send.go` and is the most important behavioral distinction in the PR.

**Gaps worth noting:**
- `--blocks` + long `text` combination in `runSend`: validation fires for non-file paths even when blocks are provided. Not tested, but the behavior is correct by inspection since `validateMessageLength` is called before the block path diverges. Low risk, minor coverage gap.
- `runSend` with long text piped via stdin (`text == "-"`): stdin is read before validation runs, so an oversized stdin message will also be rejected. No test for this path, but it's the same code path as direct text — still low risk.
- `runUpdate` has no file-upload path, so no bypass case to test. The asymmetry is correct.
- Error message includes actual character count (`%d characters`) but `TestValidateMessageLength` only checks for the static limit string, not the dynamic count. Fine for this feature, but a small assertion gap if the format changes.

**Verdict:** Coverage is adequate for the change. The critical happy/sad paths and the file-bypass carve-out are all tested. The gaps are minor and consistent with the rest of the test suite's style.

File: internal/cmd/messages/messages.go [hunk 1]
```diff
@@ -72,6 +73,27 @@ func unescapeShellChars(s string) string {
...
 const maxSectionTextLen = 3000
 
+// maxMessageTextLen is Slack's hard limit for the text field of a message.
+// Messages exceeding this limit are silently truncated by Slack.
+// See: https://docs.slack.dev/changelog/2018-truncating-really-long-messages/
+const maxMessageTextLen = 40000
+
+// validateMessageLength checks whether text exceeds Slack's message length limit.
+func validateMessageLength(text string) error {
+	if len(text) <= maxMessageTextLen {
```
Comment (monit-reviewer) [inline internal/cmd/messages/messages.go]: 🟡 **Medium** (security:code-auditor): `len(text)` counts UTF-8 bytes, not Unicode characters. Slack's 40,000 limit is on characters. A message with multi-byte characters (emoji, CJK, accented text) will be incorrectly rejected — e.g., 20,000 emoji (4 bytes each) = 80,000 bytes but only 20,000 characters, well under the limit. Use `utf8.RuneCountInString(text)` instead. Also update line 93 (`len(text)` in the error message) to report rune count so the number matches the limit.

File: internal/cmd/messages/messages_test.go [hunk 2]
```diff
@@ -1595,3 +1595,89 @@ func TestRunSend_FileUpload_NonexistentFile(t *testing.T) {
...
+		{"empty", 0, false},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			text := strings.Repeat("x", tt.length)
+			err := validateMessageLength(text)
+			if tt.wantErr {
+				require.Error(t, err)
+				assert.Contains(t, err.Error(), "exceeds Slack's 40000 character limit")
```
Comment (monit-reviewer) [inline internal/cmd/messages/messages_test.go:1620]: 🔵 **Low** (database:sql-reviewer): All test strings use ASCII-only `strings.Repeat("x", n)`, so byte count equals rune count and the `len()` vs rune-count bug is not exercised. Add a test case with multi-byte characters (e.g., `strings.Repeat("😀", maxMessageTextLen-1)`) to catch the discrepancy.

File: internal/cmd/messages/update.go [hunk 3]
```diff
@@ -43,6 +43,10 @@ func runUpdate(channel, timestamp, text string, opts *updateOptions, c *client.C
 	// Unescape shell-escaped characters (e.g., \! from zsh)
 	text = unescapeShellChars(text)
 
+	if err := validateMessageLength(text); err != nil {
```
Comment (monit-reviewer) [inline internal/cmd/messages/update.go]: 🔵 **Low** (security:code-auditor): The `update` command calls `validateMessageLength` which suggests `--file <path>` as an alternative, but `update` doesn't support `--file`. Consider using a variant error message for update, or at minimum note that `--file` applies to `send` only.

**Review by monit-reviewer — CHANGES_REQUESTED** [review]
## Automated PR Review



**Reviewed commit:** `8997bc0`

### Summary

| Reviewer | Findings |
|----------|----------|
| database:sql-reviewer | 1 |
| security:code-auditor | 2 |

<details>
<summary><strong>database:sql-reviewer (1 findings)</strong></summary>

**💡 Suggestion** - `internal/cmd/messages/messages_test.go:1618`
> All test strings use ASCII-only `strings.Repeat("x", n)`, so byte count equals rune count and the `len()` vs rune-count bug is not exercised. Add a test case with multi-byte characters (e.g., `strings.Repeat("😀", maxMessageTextLen-1)`) to catch the discrepancy.

</details>

<details>
<summary><strong>security:code-auditor (2 findings)</strong></summary>

**⚠️ Should Fix** - `internal/cmd/messages/messages.go:83`
> `len(text)` counts UTF-8 bytes, not Unicode characters. Slack's 40,000 limit is on characters. A message with multi-byte characters (emoji, CJK, accented text) will be incorrectly rejected — e.g., 20,000 emoji (4 bytes each) = 80,000 bytes but only 20,000 characters, well under the limit. Use `utf8.RuneCountInString(text)` instead. Also update line 93 (`len(text)` in the error message) to report rune count so the number matches the limit.

**💡 Suggestion** - `internal/cmd/messages/update.go:46`
> The `update` command calls `validateMessageLength` which suggests `--file <path>` as an alternative, but `update` doesn't support `--file`. Consider using a variant error message for update, or at minimum note that `--file` applies to `send` only.

</details>

*1 PR discussion thread considered.*

---
<details>
<summary>Completed in 5m 07s | $0.68</summary>

| Field | Value |
|---|---|
| Reviewers | [hybrid-synthesis](https://github.com/SignalFT/pr-review-daemon/blob/main/pr_review_daemon/reviewers/prompt.py), [database:sql-reviewer](https://github.com/SignalFT/monit-claude-plugins/tree/main/plugins/database/agents/sql-reviewer), [security:code-auditor](https://github.com/SignalFT/monit-claude-plugins/tree/main/plugins/security/agents/code-auditor), [harness-engineering:knowledge-reviewer](https://github.com/SignalFT/monit-claude-plugins/tree/main/plugins/harness-engineering/agents/knowledge-reviewer), [harness-engineering:enforcement-reviewer](https://github.com/SignalFT/monit-claude-plugins/tree/main/plugins/harness-engineering/agents/enforcement-reviewer), [harness-engineering:architecture-reviewer](https://github.com/SignalFT/monit-claude-plugins/tree/main/plugins/harness-engineering/agents/architecture-reviewer), [harness-engineering:legibility-reviewer](https://github.com/SignalFT/monit-claude-plugins/tree/main/plugins/harness-engineering/agents/legibility-reviewer) |
| Reviewed by | [pr-review-daemon](https://github.com/SignalFT/pr-review-daemon) · [monit-pr-reviewer](https://github.com/apps/monit-pr-reviewer) |
| Duration | 5m 07s (Reviewers: 1m 13s · Synthesis: 52s) |
| Cost | $0.68 |
| Tokens | 198.9k in / 19.1k out |
| Turns | 17 |

</details>

**Comment by rianjs** [comment]
All three review findings addressed in cfe1285:

1. **UTF-8 bytes vs runes** — switched to `utf8.RuneCountInString(text)` for both validation and error message reporting
2. **Multi-byte test cases** — added emoji test cases (`strings.Repeat("😀", n)`) for under/at/over limit boundaries
3. **Update error message** — `validateMessageLength` now takes a `forUpdate bool` param; update path omits the `--file` suggestion since `update` doesn't support it. Added `TestValidateMessageLength_UpdateOmitsFileHint` to verify.

**Review by monit-reviewer — APPROVED** [review]
## Automated PR Review



**Reviewed commit:** `cfe1285` | **Previous:** `8997bc0` (incremental)

### Summary

No issues found.

*5 PR discussion threads considered.*

---
<details>
<summary>Completed in 5m 18s | $0.73</summary>

| Field | Value |
|---|---|
| Mode | Re-review · Cycle 2 · Session resumed |
| Reviewers | [hybrid-synthesis](https://github.com/SignalFT/pr-review-daemon/blob/main/pr_review_daemon/reviewers/prompt.py), [database:sql-reviewer](https://github.com/SignalFT/monit-claude-plugins/tree/main/plugins/database/agents/sql-reviewer), [security:code-auditor](https://github.com/SignalFT/monit-claude-plugins/tree/main/plugins/security/agents/code-auditor), [harness-engineering:knowledge-reviewer](https://github.com/SignalFT/monit-claude-plugins/tree/main/plugins/harness-engineering/agents/knowledge-reviewer), [harness-engineering:enforcement-reviewer](https://github.com/SignalFT/monit-claude-plugins/tree/main/plugins/harness-engineering/agents/enforcement-reviewer), [harness-engineering:architecture-reviewer](https://github.com/SignalFT/monit-claude-plugins/tree/main/plugins/harness-engineering/agents/architecture-reviewer), [harness-engineering:legibility-reviewer](https://github.com/SignalFT/monit-claude-plugins/tree/main/plugins/harness-engineering/agents/legibility-reviewer) |
| Reviewed by | [pr-review-daemon](https://github.com/SignalFT/pr-review-daemon) · [monit-pr-reviewer](https://github.com/apps/monit-pr-reviewer) |
| Duration | 5m 18s (Reviewers: 1m 11s · Synthesis: 35s) |
| Cost | $0.73 |
| Tokens | 239.1k in / 19.9k out |
| Turns | 15 |

</details>
